osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/application/services/DeviceKeyProcessingService.java

```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright 2021 Alliander N.V.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ */
+package org.opensmartgridplatform.adapter.protocol.dlms.application.services;
+
+import java.time.Duration;
+import java.time.Instant;
+import java.util.Date;
+import org.opensmartgridplatform.adapter.protocol.dlms.domain.entities.DlmsDevice;
+import org.opensmartgridplatform.adapter.protocol.dlms.domain.repositories.DlmsDeviceRepository;
+import org.opensmartgridplatform.adapter.protocol.dlms.exceptions.DeviceKeyProcessAlreadyRunningException;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.stereotype.Component;
+
+@Component
+public class DeviceKeyProcessingService {
+
+  @Value("#{T(java.time.Duration).parse('${device.key.processing.timeout:PT5M}')}")
+  private Duration deviceKeyProcessingTimeout;
+
+  private final DlmsDeviceRepository dlmsDeviceRepository;
+
+  @Autowired
+  public DeviceKeyProcessingService(final DlmsDeviceRepository dlmsDeviceRepository) {
+    this.dlmsDeviceRepository = dlmsDeviceRepository;
+  }
+
+  public void startProcessing(final String deviceIdentification)
+      throws DeviceKeyProcessAlreadyRunningException {
+
+    final DlmsDevice dlmsDevice =
+        this.dlmsDeviceRepository.findByDeviceIdentification(deviceIdentification);
+    final Date keyProcessingStartTime = dlmsDevice.getKeyProcessingStartTime();
+    final Instant timeoutTime = Instant.now().minus(this.deviceKeyProcessingTimeout);
```

Comment:
It could be just me, but the naming looks confusing.
What I would intuitively understand timeoutTime to mean would be someting like the time at which the previous key processing times out (is considered to have been abandoned). That would be `keyProcessingStartTime` plus the `deviceKeyProcessingTimeout`.
If you believe the current name makes (more) sense as it is, it might be OK, but may be helped by some comments to avoid the confusion I am experiencing.

Reply:
Maybe a bit long variable name but states exactly what it is: `timeThatRunningKeyProcessShouldBeFinished`

Reply:
I think this should be less confusing

```
    ...
    final boolean processIsAlreadyRunning =
        this.hasStartedKeyProcess(dlmsDevice) && this.keyProcessBeforeTimeout(dlmsDevice);
    ...

  private boolean hasStartedKeyProcess(final DlmsDevice dlmsDevice) {
    return dlmsDevice.getKeyProcessingStartTime() != null;
  }

  private boolean keyProcessBeforeTimeout(final DlmsDevice dlmsDevice) {
    final Instant keyProcessingStartTime = dlmsDevice.getKeyProcessingStartTime();
    final Instant timeThatRunningKeyProcessShouldBeFinished =
        keyProcessingStartTime.plus(this.deviceKeyProcessingTimeout);
    return Instant.now().isBefore(timeThatRunningKeyProcessShouldBeFinished);
  }
```

Reply:
timeoutTime --> oldestStartTimeNotConsiderTimedOut